### DISCUSSION TITLE
[SDK] Unable to use Paymaster on ZkSync Sepolia Testnet using zksync-ethers (Javascript V6)

### DISCUSSION THREAD

**Discussion by ken-cronos** [zkSync-Community-Hub/zksync-developers#257] [SDKs] [answered]
### Environment

Testnet

### Acknowledgement

- [X] I understand that issues are expected with the Go SDK as they are being updated.

### Issue Description

Building a NextJS front end similar to the "hello world" front end project of the documentation.  
What's changed is that I am using zksync-ethers (Ethers V6 instead of V5 and Sepolia as the testnet (instead of Goerli).  
When using Paymaster to pay for gas fees, I am facing an error. The code works when paying gas in ETH, but does not work when using a custom ERC20 token deployed to zkSync Sepolia Testnet to pay for gas. Any suggestion welcome.

Error in browser's javascript console:

```
errors.js:122 Uncaught (in promise) TypeError: unsupported addressable value (argument="target", value=null, code=INVALID_ARGUMENT, version=6.9.1)
    at makeError (errors.js:122:21)
    at assert (errors.js:149:15)
    at assertArgument (errors.js:161:5)
    at Object.resolveAddress (checks.js:117:35)
    at Signer.sendTransaction (signer.js:97:48)
    at async handleSubmitMessageButton (page.tsx:269:33)
```

### Expected Behavior

Transaction should complete with custom token to pay for gas. (The code works when paying gas in ETH.)

### Code Example

```
import {
    BrowserProvider as ZkSyncBrowserProvider,
    Contract as zkSyncContract,
    Provider as ZkSyncProvider,
    utils as ZkSyncUtils
} from "zksync-ethers";
import {ethers} from "ethers";
...

const handleSubmitMessageButton = async (message: string | null) => {  
    if (message) {  
        const serverZkStackProvider = new ZkSyncProvider(process.env.NEXT_PUBLIC_BLOCKCHAIN_URL);  
        const userZkStackProvider = new ZkSyncBrowserProvider(window.ethereum);  
        const userZkStackSigner = await userZkStackProvider.getSigner();  
        const greeterContract = new zkSyncContract(  
            process.env.NEXT_PUBLIC_GREETER_CONTRACT_ADDRESS ? process.env.NEXT_PUBLIC_GREETER_CONTRACT_ADDRESS : "",  
            greeterContractJson.abi,  
            userZkStackSigner  
        )  
        // Configure overrides  
        let overrides = {}  
        if (selectedToken.name != "Ether") {  
            console.log("Trying to send transaction via paymaster...")  
            const testnetPaymaster =  
                await serverZkStackProvider.getTestnetPaymasterAddress();  
            if (testnetPaymaster) {  
                const gasPrice = await userZkStackProvider.getGasPrice();  
                // estimate gasLimit via paymaster  
                const initialPaymasterInput = {  
                    type: "ApprovalBased",  
                    minimalAllowance: BigInt(1),  
                    token: selectedToken.address,  
                    innerInput: new Uint8Array(),  
                }  
                const paramsForFeeEstimation = ZkSyncUtils.getPaymasterParams(  
                    testnetPaymaster,  
                    {  
                        type: "ApprovalBased",  
                        minimalAllowance: BigInt(1),  
                        token: selectedToken.address,  
                        innerInput: new Uint8Array(),  
                    }  
                );  
                // estimate gasLimit via paymaster  
                const gasLimit = await greeterContract.setGreeting.estimateGas({  
                    message,  
                    customData: {  
                        gasPerPubdata: ZkSyncUtils.DEFAULT_GAS_PER_PUBDATA_LIMIT,  
                        paymasterParams: paramsForFeeEstimation,  
                    },  
                })  
                const fee = gasPrice * BigInt(gasLimit.toString());  
                const newPaymasterInput = {  
                    type: "ApprovalBased",  
                    token: selectedToken.address,  
                    minimalAllowance: fee,  
                    // empty bytes as testnet paymaster does not use innerInput  
                    innerInput: new Uint8Array(),  
                }  
                const paymasterParams = ZkSyncUtils.getPaymasterParams(testnetPaymaster, {  
                    type: "ApprovalBased",  
                    token: selectedToken.address,  
                    minimalAllowance: fee,  
                    // empty bytes as testnet paymaster does not use innerInput  
                    innerInput: new Uint8Array(),  
                });  
                overrides = {  
                    maxFeePerGas: gasPrice,  
                    maxPriorityFeePerGas: BigInt(0),  
                    gasLimit: gasLimit,  
                    customData: {  
                        gasPerPubdata: ZkSyncUtils.DEFAULT_GAS_PER_PUBDATA_LIMIT,  
                        paymasterParams,  
                    },  
                };  
            }  
        }  
        // Execute transaction  
        const txPayload = await greeterContract.setGreeting.populateTransaction(message);
        const transaction = await userZkStackSigner.sendTransaction({...txPayload, ...overrides});  
    }  
}
```

**Comment by MexicanAce** [comment, marked as answer]
Hey @ken-cronos 👋 !

We're able to get a local version of `zksync-ethers` (Ethers V6) working with a paymaster in our Sepolia Testnet [See example here](https://github.com/dutterbutter/simple-zksync-ethers-example/tree/main). Could you share more information about your ZK Stack setup? Perhaps there's something else going on that would help provide more information.

**Reply by MexicanAce** [reply]
> instead of Goerli

Do you know if the same contracts/scripts were successful in deployment for Goerli Testnet?

**Reply by hinataa01** [reply]
test

**Reply by ken-cronos** [reply]
thank you, solved

**Comment by crimuntilve** [comment]
It seems like the issue might be related to the address format or handling in the zkSyncProvider. The error message indicates that an unsupported addressable value is being passed.

I would recommend checking the format of the token address being passed to the zkSyncProvider and ensuring that it is in the correct format expected by the library. Additionally, you might want to verify that the token contract address is correctly deployed on the zkSync Sepolia testnet and accessible by the zkSyncProvider.

You could also consider reaching out to the zkSync community or checking their documentation for any specific requirements or considerations when interacting with ERC20 tokens on zkSync.

Lastly, double-checking the version compatibility and any recent updates or changes in the zksync-ethers library could also provide insights into resolving the issue.

**Comment by sayloonkenneth** [comment]
Issue: Error encountered when attempting to use a custom ERC20 token deployed on zkSync Sepolia Testnet to pay for gas fees via Paymaster, while the code works when paying gas fees in ETH.

Solution: Check if the ERC20 token address and other parameters are correctly configured in the transaction overrides when using the custom token to pay for gas fees. Ensure that the selected token's address is properly resolved and formatted for the transaction.

**Comment by cbaca9** [comment]
The error "unsupported addressable value" indicates that there is an issue with the address value being used in the transaction. 
Ensure that the address for the custom ERC20 token deployed on zkSync Sepolia Testnet is correctly specified and compatible with the zkSyncProvider and signer being used.
